### PR TITLE
freshBootstrapTools: remove overlay, prune tools, use tar.xz archive

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes
- unpin LLVM11. fix discrepancy with freshBootstrapTools and the tools
  built on hydra. pinning the stdenv for the hydra build doesn't pin the 
  tools as the included packages are able to change.

- remove unused LLVM tools & libs which reduces the uncompressed and 
  compressed file sizes by more than 1/2. compressed tarball is now 40M 
  and uncompressed is around 200M. [list of accessed bootstrap files](https://gist.github.com/a-n-n-a-l-e-e/e69c5bb228b7efd59080c08a4c4cd423)

- remove CoreFoundation and libobjc stubs as they are not required by the bootstrap

- add `@loader_path/`. to dylibs that reference other libs in the archive.
  this is needed for libraries with re-exports.

- validate shared objects with `@rpath` references contain the reference
  in lib 

- add a test to verify that the `@loader_path/` works for libc++ as it
  re-exports libc++abi

- https://github.com/NixOS/nixpkgs/pull/295558 will incorporate the bootstrap archive

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (wegank) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:284]: Does it go against the spirit of #202347 and the unmerged https://github.com/stephank/nixpkgs/commit/7be022c2f9e9a68c07cefb803612393d7b6849db, of providing a single file in the next bootstrap files bump?

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:284]: maybe!
is the unpacked nar archive writable? i thought that it would be readonly and there are a few minor fixups that take place.
- updating the library name for dylibs
- adding scripts for egrep, fgrep, etc that want a correct path in the shebang

it looks like @stephank was able to write to the archive. thanks for the link, i will try. probably save 1-2MB and a few lines of code and is also slicker.

Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:284]: I guess so after `cp -R $tools $out && chmod -R u+w $out` in the commit.

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:284]: ah, in that case it doesn't seem like a win. the uncompressed archive is like 450 mb. so when fetchurl unpacks for us we then have to copy the uncompressed data (450MB) and then perform updates.

providing the 4 tools (bash, cpio, xz, mkdir) we save 450 MB of disk and an extra copy at the expense of needing to provide tools with no references into the nix store.

**Comment by a-n-n-a-l-e-e** [comment]
@ofborg eval

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:284]: Indeed, so `unpack = true;` should be dropped from that commit.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: This unpins the LLVM used to generate the bootstrap-tools altogether. That means you'll end up with a different bootstrap-tools as soon as Nixpkgs bumps the default LLVM version. This means rebuilding the world on Darwin and the churn this causes too much churn for Darwin Hydra.

I'm fine with bumping the LLVM version used in the bootstrap-tools but I disagree with unpinning it altogether.

**Review by toonn — COMMENTED** [review]
I can review in more detail later but I noticed a fairly important change to address first.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: Wait, isn't rebuilding the world expected when one bumps the default LLVM version on darwin?

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: i just tried:
`nix-build  pkgs/top-level/release.nix -A stdenvBootstrapTools.x86_64-apple-darwin`
after running freshbuildtools and that performed no rebuilds. so i think i am building this correctly.

it also seems that the bootstrap-tools used in the actual bootstrap haven't been updated in a long time. 2020, 2021. i just don't see what the benefit of pinning the compiler. at the very least it makes it harder to test if someone is making compiler changes thinking that just building freshBootstrapTools will test the pinned version, not realizing that the overrides is passing in an empty `[]` and thus not building with the pinned compiler. commit the change and then realize that the tools are all broken on staging-next.

Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: Yes, there should be no rebuild in this PR, since there is no rebootstrap (as in #209462). In the same spirit, pinning to `llvmPackages_16` would not help when bumping the default LLVM version to 17/18.

**Comment by a-n-n-a-l-e-e** [comment]
aarch64 build [fails](https://logs.ofborg.org/?key=nixos/nixpkgs.295557&attempt_id=ce8f9618-c02a-4560-8acd-b83c9490dc47)

```
error: output '/nix/store/rh4fmn3k232k3j4vmxnr5fxkd4qpd8g3-stdenv-bootstrap-tools' is not allowed to refer to the following paths:
         /nix/store/cii11dfrz126yb3d03aqyahwq3fanl1g-libobjc-11.0.0
```
perhaps i shouldn't have deleted these lines:
```
sed -i -e 's|/nix/store/.*/libobjc.A.dylib|@executable_path/../libobjc.A.dylib|g' \
          $out/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation.tbd
```

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: > `freshBootstrapTools` is supposed to be an easy way to build the bootstrap-tools as currently used

No, this could never be the case, since bootstrap files also contain `pcre`, `zlib`, `bzip2`, `mpfr`, etc., which could be bumped more often. Even for GCC on Linux,

- The `x86_64-linux` bootstrap files have GCC 9, before the refresh last month.
- The `powerpc64le-linux` bootstrap files on master have GCC 10, while `freshBootstrapTools` gives GCC 13.
- The `aarch64-linux` ones have GCC 12 (and are built by me!).

Instead, I consider `freshBootstrapTools` as an easy way to **refresh** the bootstrap tools.

Comment (toonn) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: You're right, @wegank. I guess I'm trying to justify how it used to be historically.

This does highlight to me that it's currently highly inconvenient to try to reproduce the bootstrap-tools. Surely we want that to be more transparent?

Comment (wegank) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: Certainly. It is easy to reproduce the bootstrap files for `x86_64-linux`, by

```bash
$ nix build github:NixOS/nixpkgs/<revision>#freshBootstrapTools.bootstrapFiles.{bootstrapTools,busybox} --system x86_64-linux
```

with the revision specified in the metadata:

https://github.com/NixOS/nixpkgs/blob/ccab9e21200bf766affe161ece719441c35e82a3/pkgs/stdenv/linux/bootstrap-files/x86_64-unknown-linux-gnu.nix#L1-L10

We can note down the same thing for the next Darwin bootstrap files update, too.

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:18]: FWIW: https://github.com/NixOS/nixpkgs/pull/295767  workaround to fix stdenvBootstrapTools on staging-next.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:287]: one option would be to use a nar file that contains the compressed cpio archive and the 4 tools needed to extract and decompress the archive. This is not as nice as having the nix layer do all the work for us, but simplifies the bootstrap-tools download so it just requires one file rather than 5.

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix]: ok -- i used a nar container to hold the compressed cpio archive and the exes needed to unpack it. 

(from PR https://github.com/NixOS/nixpkgs/pull/295558)

`bootstrapFiles` is set to the result of `<nix/fetchurl.nix>`
https://github.com/NixOS/nixpkgs/blob/30c4faca8d35a08bdf4c15ec9195867ce2b13310/pkgs/stdenv/darwin/bootstrap-files/x86_64-apple-darwin.nix#L1-L5

setup the derivation. PATH points to the exes needed to unpack and the path to the unpacked bin/
https://github.com/NixOS/nixpkgs/blob/30c4faca8d35a08bdf4c15ec9195867ce2b13310/pkgs/stdenv/darwin/default.nix#L51-L60

**Comment by a-n-n-a-l-e-e** [comment]
cc @trofi for refresh-tarballs.bash edits. tested by setting `$outpath` to local build rather than using hydra as the files will not exist until this change is merged.

<details><summary>sample output</summary>

```nix
# Autogenerated by maintainers/scripts/bootstrap-files/refresh-tarballs.bash as:
# $ ./refresh-tarballs.bash --targets=x86_64-apple-darwin
#
# Metadata:
# - nixpkgs revision: a343533bccc62400e8a9560423486a3b6c11a23b
# - hydra build: https://hydra.nixos.org/job/nixpkgs/trunk/stdenvBootstrapTools.x86_64-apple-darwin.build/latest
# - resolved hydra build: https://hydra.nixos.org/build/252571090
# - instantiated derivation: /nix/store/hdk8xb0jx47295sxk3b89jv49h5chva4-stdenv-bootstrap-tools.drv
# - output directory: /nix/store/dkd06x1idgx3a8q7qc4rw28ds403gkzn-stdenv-bootstrap-tools
# - build time: Fri, 08 Mar 2024 14:27:32 +0000
{
  bootstrapTools = import <nix/fetchurl.nix> {
    url = "http://tarballs.nixos.org/stdenv/x86_64-apple-darwin/a343533bccc62400e8a9560423486a3b6c11a23b/bootstrap-tools.tar.xz";
    hash = "sha256-PzrSLI2VSgplcw5MvYiA1hqglC5imDwInXtSy9zZKIY=";
  };
  unpack = import <nix/fetchurl.nix> {
    url = "http://tarballs.nixos.org/stdenv/x86_64-apple-darwin/a343533bccc62400e8a9560423486a3b6c11a23b/unpack.nar.xz";
    hash = "sha256-43WE2K5VsiGvibJnS1wjtwRbswdIqiezyq2KBsCcJT0=";
    unpack = true;
  };
}
```
</details>

File: maintainers/scripts/bootstrap-files/refresh-tarballs.bash
Comment (a-n-n-a-l-e-e) [inline maintainers/scripts/bootstrap-files/refresh-tarballs.bash:253]: is it possible to use `builtins.fetchTarball` here? The [docs](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-fetchTarball) say something about caching the tarball in .cache/nix/tarballs which makes me think not.

**Review by trofi — COMMENTED** [review]
`refresh-tarballs.bash` changes look good to me. Thank you!

**Review by trofi — APPROVED** [review]

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:145]: Dropping ICU is intentional right? Do believe it's fine because we're dropping it from the source releases altogether.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:230]: Cane we check in a built bootstrap-tools whether or not this is the case?

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:324]: Is there a reason for preferring sed here?

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:243]: Is this no longer the case?

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:250]: Can you explain why we can't have everything in a single NAR? Nix unpacking everything for us seems convenient.

**Review by toonn — COMMENTED** [review]
How often is `refresh-tarballs.bash` run? I'm hesitant about causing stdenv rebuilds on Darwin more often because CI already has such a hard time keeping up.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:145]: yes. it is intentional. i looked at the access times from the bootstrap-tools folder and it was not used so i dropped it.
https://gist.github.com/a-n-n-a-l-e-e/e69c5bb228b7efd59080c08a4c4cd423

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:230]: this is an artifact and was introduced in a previous change. i added the comment as it looks wrong but i haven't written test code to figure out what the deal is with tbd files. suggestions are welcome.

do i just link to a tbd file like i would a dylib?  i think yes.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:324]: so i moved from `stdenv.mkDerivation` to a plain `derivation` to not get a polluted environment to test and there is not a substitueInPlace available in `derivation`.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:243]: forgot to delete -- it's gone now.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:250]: https://github.com/NixOS/nixpkgs/pull/295557#discussion_r1522944429

using a single NAR doesn't seem like it buys us much as we still need to patch the uncompressed output. if nix uncompressed it is 200MB read-only in the nix store which we then need to copy and patch. if we do the expand step it doesn't wast 200MB and a 200MB copy, so i didn't think that putting everything in the NAR was a win.

there are only a few tools needed to expand the tar file and they aren't hard to pack.

**Comment by a-n-n-a-l-e-e** [comment]
> How often is `refresh-tarballs.bash` run? I'm hesitant about causing stdenv rebuilds on Darwin more often because CI already has such a hard time keeping up.

i think `refresh-tarballs.bash` is run manually. i think that this change will reduce pressure on hydra as the stdenvBootstrapTools job will not have to build all the tools under the llvm11 stdenv but can use the cached ones already built with the default stdenv.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:230]: from some trial and error. when linking to `CoreFoundation.tbd` and modifying the contents of `libraries: ['@executable_path/..//llibobjc.A.dylib']`
- `@executable_path` represents the current directory of the linker
- `@loader_path` represents the current directory of the CoreFoundation.tbd
- if the basename is `libobjc.A.dylib` it doesn't matter what the parent is, link will succeed
- renaming the `libobjc.A.tbd` -> `llibobj.A.tbd` will create linker errors if path is not correct

so using `@loader_path` seems like it is correct except if the file is symlinked then `@loader_path` is relative to the symlink and not the target. which breaks. this is not the case for rpaths in dylibs as `@loader_path` through a symlink always seems to be the location of the target.

anyway, it seems like these files need to be patched with an absolute path.

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix]: fixed to patch in absolute path. verified path correct in aarch64.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (reckenrode) [inline pkgs/stdenv/darwin/make-bootstrap-tools.nix:148]: With both platforms no longer using the source-based CF, can this be dropped? It’s just the stubs now.  With both `pbzx` and `xar` being included, it should be possible to build CF in stage 1 of the stdenv bootstrap.

(I’m less certain about Libsystem since it comes with object files on x86_64-darlin. It might be worth keeping, and if the bootstrap can be made not to use it, then it can be dropped for the next bootstrap tools update.)

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:148]: the CF headers from boostrapTools are still accessed: https://gist.github.com/a-n-n-a-l-e-e/e69c5bb228b7efd59080c08a4c4cd423

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix]: ok -- was able to remove `pkgs.darwin.CF` and `pkgs.darwin.libobjc`. ran aarhc64 and x64 bootstraps to stage2. updated the tests to remove the CF test.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix]: was able to remove CF as suggested by @reckenrode which removes the file from the bootstrap tools.

**Comment by a-n-n-a-l-e-e** [comment]
moved the unpack/patch script into the archive `unpackScript` so it is possible to make edits to the script and and test without triggering a stdenv rebuild.

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (toonn) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:288]: The fixups you mention, why do those need absolute paths? I assume that's the reason they can only happen after unpacking?

**Comment by toonn** [comment]
My question re the frequency of `refresh-tarballs.bash` was based on the mistaken assumption that it's run regularly for Linux and when run would always run for Darwin too but it seems the target platform is explicitly specified?

File: pkgs/stdenv/darwin/make-bootstrap-tools.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/darwin/make-bootstrap-tools.nix:288]: so this is due running the following on each dylib
```
          install_name_tool -id "$path" "$path.new"
          codesign -f -i "$(basename "$path")" -s - "$path.new"
```
where `$path` is the absolute path to the dylib. this code predates my changes and i do not know why an absolute path is needed.

there are a couple of shell scripts added that use full path shebangs but that can probably be worked around.

**Comment by a-n-n-a-l-e-e** [comment]
I think all the issues are resolved and i'd like to move forward in getting the tools built on hydra so i can do the second part of this change to use the new tools. the code has been in an inconsistent state for over a year where it was not possible, without changes,  to use newly built tools for the bootstrap.

i'll let this sit for another 24 hours or so and if no objections will merge. thanks for all the reviews and matrix discussions!

**Comment by vcunat** [comment]
@ofborg build freshBootstrapTools.test

**Review by toonn — APPROVED** [review]
LGTM
Are we just waiting on the tests to pass?

**Review by wegank — APPROVED** [review]
Probably for the 24 hours' notice.

**Comment by vcunat** [comment]
`aarch64-darwin` on ofBorg tends to be too overloaded, surely not worth waiting for.

**Comment by a-n-n-a-l-e-e** [comment]
the last commit was just a rebase with no changes and the ofborg tests passed prior -- and tests passed on the aarch64 community builder -- so wasn't planning on waiting for test to finish. just didn't want to merge without approvals without giving ample notice -- but now there are approvals seems fine.

thanks for the reviews!

**Pull request merged**
